### ISSUE TITLE
jrnl uses UTC instead of local time for entries in WSL/Ubuntu

### ISSUE THREAD

**Issue by micahellison** [jrnl-org/jrnl#1607]
### Diagnostic output

```
jrnl: v3.2
Python: 3.10.7 (main, Sep  5 2022, 13:12:31) [GCC 11.2.0]
OS: Linux 4.4.0-19041-Microsoft
```

### Current Behavior

The timestamp associated with new entries uses UTC instead of local time.

### Expected Behavior

jrnl is expected to use local time.

### Repro Steps

```
$ jrnl I am making a new entry at 7:26 AM
$ jrnl -1
2022-10-06 14:26 I am making a new entry at 7:26 AM
```

Note that the entry time is 14:26 instead 07:26.

Checking the system date immediately afterward:
```
$ date
Thu Oct  6 07:27:34 PDT 2022
```

### Debug output

```
$ jrnl --debug jrnl I am making a new entry at 7:33 AM
DEBUG    root         Parsed args: Namespace(debug=True, preconfig_cmd=None, postconfig_cmd=None, filename=None, on_date=None, today_in_history=False, month=None, day=None, year=None, start_date=None, end_date=None, contains=None, strict=False, starred=False, limit=None, excluded=[], edit=False, delete=False, change_time=None, export=False, tags=False, short=False, config_override=[], config_file_path='', text=['jrnl', 'I', 'am', 'making', 'a', 'new', 'entry', 'at', '7:33', 'AM'])
DEBUG    root         Reading configuration from file /home/*******/.config/jrnl/jrnl.yaml
DEBUG    root         Using configuration "{'colors': {'body': 'none', 'date': 'none', 'tags': 'none', 'title': 'none'}, 'default_hour': 9, 'default_minute': 0, 'editor': '', 'encrypt': False, 'highlight': True, 'indent_character': '|', 'journals': {'default': '/home/*******/.local/share/jrnl/journal.txt'}, 'linewrap': 79, 'tagsymbols': '#@', 'template': False, 'timeformat': '%Y-%m-%d %H:%M', 'version': 'v3.0'}"
DEBUG    root         Using journal name: default
DEBUG    root         opened PlainJournal with 16 entries
DEBUG    root         Write mode: starting
DEBUG    root         Write mode: cli text detected: ['jrnl', 'I', 'am', 'making', 'a', 'new', 'entry', 'at', '7:33', 'AM']
DEBUG    root         Write mode: appending raw text to journal "default": jrnl I am making a new entry at 7:33 AM
DEBUG    root         Write mode: completed journal.write()
```

### Other Information

My config is the default config. Timeformat is: `'%Y-%m-%d %H:%M'`

~~Interestingly enough, queries like -from are working just fine across days regardless of timestamp. I suspect it's just a problem with how the data is saved.~~ (this is wrong, see my comment below)

I don't know if the problem is specific to the combination of WSL/Ubuntu, though I am not running into it using jrnl on the same Windows 10 Pro machine through Windows directly (via manual Python install & pipx) instead of WSL/Ubuntu.

This issue first reported in discussion #1606. Thanks @jordanstudioroot for raising it.

**Comment by micahellison** [comment]
It turns out this issue is happening in searches too.

For instance, if I have an entry saved at 19:20pm and I run this at 12:20pm PST:
```
$ jrnl -from "five minutes ago" -to "five minutes from now"
```
It will display the 19:20pm entry.

**Comment by AlexBeez** [comment]
Issue confirmed to happen on Peppermint/Devuan as well.

**Comment by giuseppedandrea** [comment]
After a deep analysis, I'm quite sure to have found the source of the issue and it is not a bug related to jrnl. Sorry for the long comment, possible workarounds are explained at the end.

When a new entry is created (`new_entry` method of `Journal` class), the raw text from the user is received and there are two possibilities:

- If a date/time is specified in the first line of the raw text (e.g. "yesterday at 2 am: Title..."), `time.parse` function from the `jrnl` module is called passing the string before the colon
- If a date/time is not specified in the first line of the raw text, `time.parse` function from the `jrnl` module is called passing the "now" string

The `time.parse` function from the `jrnl` module (which is also called to filter entries with search command) returns a `datetime` object built using the `dateutil.parser` module (which parses known formats to represent a date and/or time) or the `parsedatetime` module (which parses human-readable date/time strings).

The important thing to notice is that almost all Python modules dealing with date and time (e.g. `datetime`, `parsedatetime`, `dateutil`, etc.) use internally the `time` module, where [timezone constants](https://docs.python.org/3/library/time.html#timezone-constants) are initialized.
So, I did some basic tests using the `time` module of Python on different machines with different configurations:

- Python 3.11 (installed with Python Installer) on Windows 10
  ```
  > python.exe
  Python 3.11.4 (tags/v3.11.4:d2340ef, Jun  7 2023, 05:45:37) [MSC v.1934 64 bit (AMD64)] on win32
  Type "help", "copyright", "credits" or "license" for more information.
  >>> import time
  >>> time.localtime()
  time.struct_time(tm_year=2023, tm_mon=8, tm_mday=2, tm_hour=18, tm_min=7, tm_sec=17, tm_wday=2, tm_yday=214, tm_isdst=1)
  
  > Get-Date
  mercoledì 2 agosto 2023 18:07:35
  ```
- Python 3.11 (installed using Homebrew) on Ubuntu 20.04 on WSL2
  ```
  > python3.11
  Python 3.11.4 (main, Jun  6 2023, 22:16:46) [GCC 11.3.0] on linux
  Type "help", "copyright", "credits" or "license" for more information.
  >>> import time
  >>> time.localtime()
  time.struct_time(tm_year=2023, tm_mon=8, tm_mday=2, tm_hour=16, tm_min=5, tm_sec=55, tm_wday=2, tm_yday=214, tm_isdst=0)

  > date
  Wed Aug  2 18:05:58 CEST 2023
  ```

As you can notice from the above outputs, Python installed through Homebrew on Ubuntu 20.04 on WSL2 returns the wrong timezone constants. So, the problem described in this issue seems not related to jrnl, but a more general issue with the `time` module of Python.

I started looking into the source code (written in C) of the `time` module. I'll skip all the details but the important thing to notice is that the `tzset()` function from the C standard library is responsible to initialize time conversion information (i.e. `tzset()` function is automatically called by other time conversion functions that depend on the timezone). In particular:

- The `tzset()` function initializes the `tzname` variable from the `TZ` environment variable
- If the `TZ` variable does not appear in the environment, the system timezone is used
- The system timezone is configured by copying, or linking, a file in the [tzfile format](https://man7.org/linux/man-pages/man5/tzfile.5.html) to */etc/localtime*
- If the default tzfile is not found (or it is in the wrong format), the UTC time is used

Since the `time` module is written in C, I wrote a simple Python script (`import time; tm  = time.localtime(); print(tm);`) and I started GDB on it (`gdb --args python3.11 test.py`). I set a breakpoint to the function where the tzfile is read (`__tzfile_read()`) and I noticed two things:

- From the call stack I noticed that Python was using the `glibc` version of Homebrew (*/home/linuxbrew/.linuxbrew/opt/glibc/lib/libc.so.6*)
- From the registers I found that the path passed as argument to the function was */home/linuxbrew/.linuxbrew/etc/localtime*

But this tzfile file doesn't exist on my machine. So, since the default tzfile is not found, the UTC time is used.

Actually, as reported in [this discussion](https://github.com/orgs/Homebrew/discussions/3617#discussioncomment-3648992) about a similar issue with git installed through Homebrew, the `post_install` step in `glibc` Homebrew formula [should set](https://github.com/Homebrew/homebrew-core/blob/120cd1a9de64965d1bc29a09c5942bca9f6dfc83/Formula/glibc.rb#L228-L231) the local time zone file, but it seems there is a bug when the symbolic link is created. I'll open an issue on the proper repository.

Hence, if jrnl is installed using Homebrew and if you also have `glibc` formula installed in Homebrew, Python will use this library and the time zone used is UTC because the tzfile doesn't exist and the `TZ` environment variable is not set.

There are two workarounds to this issue:

- Create manually the missing symbolic link in the Homebrew installation path
  - Command: ``ln -s /etc/localtime `brew --prefix`/etc/localtime``
- Set the `TZ` environment variable
  - Command: `export TZ=":Europe/Berlin"`
  - *Europe/Berlin* is a timezone file relative to the system timezone directory (*/usr/share/zoneinfo/*)
  - Replace *Europe/Berlin* with your timezone file

**Comment by micahellison** [comment]
This is great, @giuseppedandrea. Thank you so much for the deep dive, for raising these issues in the relevant repos, and documenting the workarounds here.

**Comment by giuseppedandrea** [comment]
@micahellison The bug in the `glibc` formula has been resolved (Homebrew/homebrew-core#138629). However, a bottle (binary package) including the fix has not been generated yet, and I don't know when it will happen.

**Comment by micahellison** [comment]
Thanks for the update and the resolution, @giuseppedandrea. I've just used brew to install jrnl v4.0.1_4 on WSL/Ubuntu and I see that it's using the appropriate time, so I think we can call this issue resolved.

I'm closing this as "won't fix" since it's been resolved outside of the jrnl repository, though if anyone runs into this issue again, please feel free to respond to this issue or open a new one.
